app/models/goods_nomenclatures/nested_set.rb

```diff
@@ -0,0 +1,129 @@
+module GoodsNomenclatures
+  module NestedSet
+    extend ActiveSupport::Concern
+
+    END_OF_TREE = 1_000_000_000_000
+
+    included do
+      one_to_one :tree_node, key: :goods_nomenclature_sid,
+                             class_name: 'GoodsNomenclatures::TreeNode',
+                             reciprocal: :goods_nomenclature,
+                             read_only: true do |ds|
+        ds.with_actual(GoodsNomenclatures::TreeNode)
+      end
+
+      many_to_many :ns_ancestors,
+                   left_primary_key: :goods_nomenclature_sid,
+                   left_key: Sequel.qualify(:origin_nodes, :goods_nomenclature_sid),
+                   right_primary_key: :goods_nomenclature_sid,
+                   right_key: :goods_nomenclature_sid,
+                   class_name: '::GoodsNomenclature',
+                   join_table: Sequel.as(:goods_nomenclature_tree_nodes, :ancestor_nodes),
+                   read_only: true do |ds|
+        ds.order(:ancestor_nodes__depth)
+          .with_validity_dates(:ancestor_nodes)
+          .select_append(:ancestor_nodes__depth)
+          .join(Sequel.as(:goods_nomenclature_tree_nodes, :origin_nodes)) do |origin_table, ancestors_table, _join_clauses|
+            ancestors_position = Sequel.qualify(ancestors_table, :position)
```

Comment:
There's quite a bit of duplication between the association methods with these variables? Is it worth extracting them out?

Reply:
Fair point - I'll have a think, the complication is there's no object to attach the state to so you just end up pass a lot of params around - should be able to do something though